cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i

```diff
@@ -148,40 +149,101 @@ services::Status KernelImplLinear<fastCSR, algorithmFPType, cpu>::computeInterna
                 dataR[i * nVectors1 + j] = dataR[j * nVectors1 + i];
             }
         });
+        return services::Status();
     }
-    else
-    {
-        ReadRowsCSR<algorithmFPType, cpu> mtA2(dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(a2)), 0, nVectors2);
-        DAAL_CHECK_BLOCK_STATUS(mtA2);
-        const algorithmFPType * dataA2 = mtA2.values();
-        const size_t * colIndicesA2    = mtA2.cols();
-        const size_t * rowOffsetsA2    = mtA2.rows();
 
-        SpBlas<algorithmFPType, cpu>::xgemm_a_bt(dataA1, colIndicesA1, rowOffsetsA1, dataA2, colIndicesA2, rowOffsetsA2, nVectors1, nVectors2,
-                                                 a1->getNumberOfColumns(), dataR);
+    const size_t blockSize = 256;
+    const size_t nBlocks1  = nVectors1 / blockSize + !!(nVectors1 % blockSize);
+    const size_t nBlocks2  = nVectors2 / blockSize + !!(nVectors2 % blockSize);
 
-        if (k != (algorithmFPType)1.0 || b != (algorithmFPType)0.0)
+    const bool isSOARes = r->getDataLayout() & NumericTableIface::soa;
+
+    TlsMem<algorithmFPType, cpu> tlsMklBuff(blockSize * blockSize);
+    SafeStatus safeStat;
+    daal::conditional_threader_for((nVectors1 >= blockSize * 2), nBlocks1, [&, isSOARes](const size_t iBlock1) {
+        const size_t nRowsInBlock1 = (iBlock1 != nBlocks1 - 1) ? blockSize : nVectors1 - iBlock1 * blockSize;
+        const size_t startRow1     = iBlock1 * blockSize;
```

Comment:
```cpp
const size_t nRowsInBlock1 = min(startRow1 + blockSize, nVectors1) - startRow1;
```

The same below.

Reply:
In many places, this is exactly how it is done, if we change it - then everywhere.